LibGL+LibSoftGPU: Implement shaders

supercomputer7 opened the pull request:
This is an attempt to revive from the work that has been done in #15272.

It had a couple of conflicts because of changes in LibGL recently. I still need to fix the last comments of that PR before this is mergeable :)

cc @sunverwerth

sunverwerth commented on Userland/Libraries/LibGL/Shaders/Shader.h:
```diff
@@ -26,6 +28,7 @@ class Shader final : public RefCounted<Shader> {
     ErrorOr<void> compile();
     GLenum type() const { return m_type; }
     bool compile_status() const { return m_compile_status; }
+    GLSL::ObjectFile const* object_file() const { return m_object_file.ptr(); }
```
Was this added in my original PR? Looks weird, maybe returning a const reference would be nicer.

supercomputer7 replied:
Not sure, I will fix this otherwise.

supercomputer7 replied:
On second thought, I don't think we need to change this. Everything else that uses this function assumes it will get a pointer, and it's not guaranteed that you will have a valid reference to the `ObjectFile`.

sunverwerth replied:
Right, I think I did it like that because of the `ObjectFile` vector in `GL::Program::link()`. I remember this was done on purpose.

GMTA commented on Userland/Libraries/LibGL/Shaders/Program.cpp:
```diff
@@ -13,9 +14,40 @@ NonnullRefPtr<Program> Program::create()
+
+    case GL_FRAGMENT_SHADER:
+        return m_fragment_shaders.contains_slow(shader);
+        break;
```
We don't need these `break`s after a `return`.

GMTA commented on Tests/LibGL/TestShaders.cpp:
```diff
@@ -0,0 +1,69 @@
+#else
+#    define REFERENCE_IMAGE_DIR "reference-images"
+#endif
+#define SAVE_OUTPUT false
```
These defines aren't used.

GMTA commented on Userland/Libraries/LibGL/Shader.cpp:
```diff
@@ -149,7 +149,7 @@ void GLContext::gl_link_program(GLuint program)
     RETURN_WITH_ERROR_IF(program_it == m_allocated_programs.end(), GL_INVALID_OPERATION);
     // FIXME: implement check "GL_INVALID_OPERATION is generated if program is the currently active program object and transform feedback mode is active."
-    
+
```
This whitespace fix should go into the commit that introduced it

GMTA commented on Userland/Libraries/LibSoftGPU/Device.cpp:
```diff
@@ -809,13 +812,14 @@ void Device::rasterize_triangle(Triangle& triangle)
+                quad.set_input(SHADER_INPUT_VERTEX_COLOR, interpolate(expand4(vertex0.color), expand4(vertex1.color), expand4(vertex2.color), quad.barycentrics));
+            } else {
+                quad.set_input(SHADER_INPUT_VERTEX_COLOR, expand4(vertex0.color));
+            }
```
These braces are not necessary for one-liners; they weren't there before

GMTA commented:
Thanks @supercomputer7 ! There are also some comments in the old PR that are still relevant, but I think you found those.

sunverwerth commented on Userland/Libraries/LibGL/Shader.cpp:
```diff
@@ -0,0 +1,209 @@
+
+    it->value->clear_sources();
+    for (int i = 0; i < count; i++) {
+        if (!length || length[i] < 0)
```
```suggestion
        if (length == nullptr || length[i] < 0)
```

Copying @Quaker762 's suggestion from the old PR.

supercomputer7 commented:
This should be ready for merging to the tree now.

sunverwerth reviewed (approved):
LGTM, but since I was the original author maybe someone else should have the final say :^) @GMTA ?

supercomputer7 commented:
Seems like CI failed because of network errors. So I think it's still mergeable.

sunverwerth commented:
I tested a few of our 3d apps on this branch. Everything still seems to be in order.
![Screenshot at 2022-12-04 14-49-29](https://user-images.githubusercontent.com/2094371/205495257-4df9b3d7-0f77-4049-bd4e-1401448f1ae7.png)
![Screenshot at 2022-12-04 15-00-14](https://user-images.githubusercontent.com/2094371/205495263-d1515f22-1e36-43f2-bf2c-73a4d2d3d4ab.png)
![Screenshot at 2022-12-04 15-03-42](https://user-images.githubusercontent.com/2094371/205495266-e83114ec-8369-45a1-bad6-3664accd9400.png)
![Screenshot at 2022-12-04 15-06-55](https://user-images.githubusercontent.com/2094371/205495269-8493ba47-3f72-4d57-b045-432df4c07295.png)

sunverwerth commented:
Reminds me that we badly need more automated tests for LibGL :^)

ADKaster commented on Userland/Libraries/LibGL/Shaders/Shader.h:
```diff
@@ -0,0 +1,49 @@
+    static NonnullRefPtr<Shader> create(GLenum shader_type);
+
+    void clear_sources() { m_sources.clear(); }
+    void add_source(StringView source_code) { m_sources.append(source_code); }
```
This code needs some help regarding the String --> DeprecatedString refactor. At least m_sources and m_info_log here should be converted to DeprecatedString entries until we can refactor later.

supercomputer7 commented:
I refactored this to use the new String class. While I looked through the code, I found places where the code could return an `Error` in a function that has `ErrorOr<void>` return type but some parts of the function neglected it, so I fixed these few cases as well.
I consider this ready to merge as soon as CI is green and @sunverwerth puts his approval on this.

sunverwerth commented on Userland/Libraries/LibGL/Shaders/Shader.cpp:
```diff
@@ -0,0 +1,30 @@
+
+ErrorOr<void> Shader::add_source(StringView source_code)
+{
+    auto source_code_name = TRY(String::from_utf8(source_code));
```
Why is this called `source_code_name`?
